### DISCUSSION TITLE
How to test code that uses Node's worker_threads

### DISCUSSION THREAD

**Discussion by chadhietala** [vitest-dev/vitest#3419] [Q&A]
What is the recommended way to refer to a file that is loaded by a worker so that it works in tests. I'm currently doing the following:

```ts
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const workerPath = resolve(__dirname, 'graphWorker.js');

//...

const worker = new Worker(workerPath);
```

Currently I get the following error which makes sense because graphWorker.js does not exist since the implementation is in TypeScript and the files are `.ts`. 

```
Error: Cannot find module '/Users/chietala/Code/rehearsal-js/packages/cli/src/commands/graph/tasks/graphWorker.js'
```

Is there a way to get Vitest to see load the worker correctly?

**Comment by cansin** [comment]
I am also trying to find an answer to this exact question.

**Comment by AriPerkkio** [comment]
This is currently not supported. You need to have the actual `.js` file on filesystem. 

I guess one could write a custom Vite plugin that captures `new Worker('<filename>.ts')` from source files. It would need to transform the original file to JS and write it on filesystem, and then replace the filename in test file. The `Worker` cannot load virtual modules so an actual file write is needed. 

It would be nice to support this in Vitest out-of-the-box.

**Reply by s-h-a-d-o-w** [reply]
I recently created a plugin that does that: https://github.com/s-h-a-d-o-w/esbuild-plugin-node-worker
However, this in the vitest config doesn't work:
```ts
  optimizeDeps: {
    esbuildOptions: {
      plugins: [nodeWorker],
    },
  },
```

**Comment by TkachenkoDmitry** [comment]
You can pass `execArgv` to worker:
```   
    const worker = new Worker('./worker.ts', {
      execArgv: ['--require', 'ts-node/register'],
      workerData: {},
    });
```
this will tell it to use `ts-node`

**Reply by AriPerkkio** [reply]
That's good work-around for now. As the code would be loaded outside Vitest, you cannot use Vitest features like mocking inside the `worker.ts`.

**Reply by s-h-a-d-o-w** [reply]
Thanks a lot for pointing out this option! Although I would recommend using the following instead, so that ts-node doesn't have to be installed:

```
execArgv: ["--experimental-transform-types"],
```

**Comment by suguanYang** [comment]
Using vite-node's programmatic API,  https://github.com/vitest-dev/vitest/tree/main/packages/vite-node#programmatic-usage

// worker-wrapper.js
```
import { workerData } from 'node:worker_threads';

import { createServer } from 'vite'
import { ViteNodeRunner } from 'vite-node/client'
import { ViteNodeServer } from 'vite-node/server'
import { installSourcemapsSupport } from 'vite-node/source-map'

// create vite server
const server = await createServer({
    optimizeDeps: {
        // It's recommended to disable deps optimization
        noDiscovery: true,
        include: undefined,
    },
    server: {
        port: 21333
    }
})

// create vite-node server
const node = new ViteNodeServer(server)

// fixes stacktraces in Errors
installSourcemapsSupport({
    getSourceMap: source => node.getSourceMap(source),
})

// create vite-node runner
const runner = new ViteNodeRunner({
    root: server.config.root,
    base: server.config.base,
    // when having the server and runner in a different context,
    // you will need to handle the communication between them
    // and pass to this function
    fetchModule(id) {
        return node.fetchModule(id)
    },
    resolveId(id, importer) {
        return node.resolveId(id, importer)
    },
})


const mod = await runner.executeFile(workerData.fullpath)

export default mod.default
```
